### PR TITLE
Fix support for symlink save_dir in TensorBoardLogger

### LINKED ISSUE #6690: symlink not accepted as logger folder
## 🐛 Bug

Moving `lightning_logs` to another filesystem and symlinking it so I don't need to touch my code, I notice that the logger rejects symlinked logger folders:

`loggers/csv_logs.py        if not os.path.isdir(root_dir):`

Users will expect symlinks to folders to work the same way as folders.

### To Reproduce

```
mkdir /tmp/lightning_logs
ln -s /tmp/lightning_logs/
```
and try training as usual.

### Expected behavior

The symlinked folder is accepted and, in the above example, the logs start filling /tmp instead of $HOME.

### Workarounds

(Breaking existing paths)

Solution 1: Create a folder inside the symlinked folder and set `save_dir` to this subfolder (not tested)
Solution 2: Use the absolute path to the folder, e.g. `trainer = pl.Trainer(logger = pl.loggers.TensorBoardLogger(os.path.abspath('lightning_logs')))` (I was expecting the next run to go to `version_3` after I had versions 0, 1 and 2 but it went to `default/version_0` so there still seems to be a side effect)

### Environment

Private, local Jupyter notebook with manually setup virtualenv.

 - PyTorch Version (e.g., 1.0): 1.7.1+cu101
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.6.12
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: GeForce GTX Titan-X
 - Any other relevant information: None

### Additional context

None

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes #6690

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/cloud_io.py [hunk 1]
```diff
@@ -12,15 +12,31 @@
...
+        try:
+            if self.info(path)["type"] == "directory":
+                return True
+            elif self.info(path)["type"] == "link":
+                return os.path.isdir(path)
+            else:
+                return False
+        except IOError:
+            return False
+
```
Comment (jowagner) [inline pytorch_lightning/utilities/cloud_io.py:38]: Suggest to only test for "link" and delegate all other cases to super().isdir()

Comment (ethanwharris) [reply pytorch_lightning/utilities/cloud_io.py:40]: @jowagner That's a great suggestion, no need for the disclaimer! All done :smiley:

Comment (ethanwharris) [reply pytorch_lightning/utilities/cloud_io.py:34]: On second thought, can actually just use `os.path.isdir` (which is what `.info` uses anyway, only it checks that first then defers to isdir otherwise).

File: pytorch_lightning/utilities/cloud_io.py [hunk 2]
```diff
@@ -12,15 +12,27 @@
...
 import torch
 
 
+class LightningLocalFileSystem(LocalFileSystem):
+    """Extension of ``fsspec.implementations.local.LocalFileSystem`` where ``LightningLocalFileSystem.isdir`` behaves
+    the same as ``os.isdir``.
+    """
+
+    def isdir(self, path: str) -> bool:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/cloud_io.py:32]: Could you explain this, why we have to provide this implementation?
I followed the code of ffsspec LocalFileSystem

https://github.com/intake/filesystem_spec/blob/891fb609767f0ddad8b24520335654ff3c13302f/fsspec/spec.py#L625

And it calls info which itself checks for symlink.

File: tests/loggers/test_tensorboard.py [hunk 3]
```diff
@@ -301,3 +301,22 @@ def test_tensorboard_save_hparams_to_yaml_once(tmpdir):
...
     assert os.path.isfile(os.path.join(trainer.log_dir, hparams_file))
     assert not os.path.isfile(os.path.join(tmpdir, hparams_file))
+
+
+@mock.patch('pytorch_lightning.loggers.tensorboard.log')
+def test_tensorboard_with_symlink(log, tmpdir):
+    """
+    Tests a specific failure case when tensorboard logger is used with empty name, symbolic link ``save_dir``, and
+    relative paths.
+    """
```
Comment (awaelchli) [inline tests/loggers/test_tensorboard.py:311]: it fails on master, right?

**Review by awaelchli — APPROVED** [review]
nice

File: tests/loggers/test_tensorboard.py (hunk 3 above)
Comment (ethanwharris) [reply tests/loggers/test_tensorboard.py:311]: Yeah, this fails on master

File: pytorch_lightning/utilities/cloud_io.py (hunk 2 above)
Comment (ethanwharris) [reply pytorch_lightning/utilities/cloud_io.py:32]: Sure, ffsspec `LocalFileSystem.info` checks for links, but the returns `type = 'link'` here: https://github.com/intake/filesystem_spec/blob/891fb609767f0ddad8b24520335654ff3c13302f/fsspec/implementations/local.py#L59

But `AbstractFileSystem.isdir` returns `true` only if `type = 'directory'`. So in this case `isdir` returns false even though it is a directory (just also a sym link).

Comment (ananthsub) [reply pytorch_lightning/utilities/cloud_io.py:32]: is this something that should be fixed in fsspec upstream?

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/utilities/cloud_io.py (hunk 2 above)
Comment (ethanwharris) [reply pytorch_lightning/utilities/cloud_io.py:32]: @ananthsub Yeah, would be better if this was fixed upstream. I have opened an issue here: https://github.com/intake/filesystem_spec/issues/591 - will remove this fixed once / if that gets adressed :smiley:

**Review by akihironitta — APPROVED** [review]

File: pytorch_lightning/utilities/cloud_io.py [hunk 4]
```diff
@@ -12,15 +12,27 @@
...
 import fsspec
+from fsspec.implementations.local import LocalFileSystem
+
 import torch
 
 
+class LightningLocalFileSystem(LocalFileSystem):
+    """Extension of ``fsspec.implementations.local.LocalFileSystem`` where ``LightningLocalFileSystem.isdir`` behaves
+    the same as ``os.isdir``.
+    """
```
Comment (carmocca) [inline pytorch_lightning/utilities/cloud_io.py:30]:
```suggestion

    To be removed when https://github.com/intake/filesystem_spec/issues/591 is fixed.
    """
```

File: pytorch_lightning/utilities/cloud_io.py [hunk 5]
```diff
@@ -12,15 +12,27 @@
...
 from pathlib import Path
 from typing import IO, Union
 
 import fsspec
+from fsspec.implementations.local import LocalFileSystem
+
 import torch
 
 
+class LightningLocalFileSystem(LocalFileSystem):
```
Comment (carmocca) [inline pytorch_lightning/utilities/cloud_io.py:27]: Can we make this protected in case this stays here for a while?

```suggestion
class _LightningLocalFileSystem(LocalFileSystem):
```

**Review by carmocca — CHANGES_REQUESTED** [review]

**Review by s-rog — APPROVED** [review]

**Comment by ethanwharris** [comment]
@carmocca Thanks for the suggestions! All done :smiley:

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
